design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,495 @@
+    It has the virtue of being simple.  If you’re making changes, you just do them directly.
+- Draw indirectly.   <br>Metal does this.  You add your data to a command buffer where it may be drawn for many frames.
+    This is the fastest and best way to do things, but there are updates you must make between frames.  Obviously, the overall map state needs to be updated, otherwise you’re just redrawing in the same way each time.  Sometimes Drawable specific state changes too and data driven rendering will come into play here.
+- Update for frame.   <br>If you’re drawing indirectly, this is all the state that must be changed for a given frame.
+    Some of this is shared, like the map state (e.g. the matrix controlling positioning, lighting and so forth).
+    Some can be specific to a particular Drawable.  Data driven visuals may fall into this category depending on what the data is.<br>
+    Just to make this even more fun, modern renderers are going to have multiple frames in flight at any given time.  So you can’t just keep one set of values and update them periodically.  My preferred approach is to put memory copy commands into a command buffer with guard logic between them.  In Metal, anyway.  Vulkan will have its own way to do that and OpenGL just doesn't.  Well, there's probably an extension somewhere that does, but it's very hard to use and sparsely supported.
+- Tear down their data.   <br>Pretty simple for OpenGL, but with Metal when you’re using heaps (and you should) you actually want another thread to do this.  Thus it’s SDK specific.
+
+Now there is already logic to do lot of this spread throughout various classes in MapLibre Native.  Buckets have some of it, Programs actually own the draw() method, and so forth.  To switch to this approach we'll need to cut across the gl and glx levels of the toolkit, even a bit higher, to capture everything that builds geometry.  We'll need to convert that over to this approach and make sure we didn't miss anything, like all the fiddly per-program state.
+
+#### Benefits
+
+Switching to a Drawable/Builder approach helps consolidate the rendering SDK specific logic in a smaller number of places.  OpenGL state is notoriously picky, Metal is less so still but still complex.  This is a case where making general purpose uber-classes is actually the right choice.
+
+We can make things like the location puck simpler.  If we can expose much of this outside the toolkit we may even make it possible to do things like your own location puck, or similar functionality.
+
+By moving to this approach we can let the rendering logic deal only with its own classes.  Rather than letting implementaton propagate all the way down to the render_*_layer, which has to very specifically implement them, we can make layers ask for what they need and let the SDK level implement it however it may feel like.  This is particularly important for Metal's indirect mode.
```

Comment:
```suggestion
By moving to this approach we can let the rendering logic deal only with its own classes.  Rather than letting implementation propagate all the way down to the render_*_layer, which has to very specifically implement them, we can make layers ask for what they need and let the SDK level implement it however it may feel like.  This is particularly important for Metal's indirect mode.
```